entity-service-impl/src/main/java/org/hypertrace/entity/query/service/converter/UpdateConverter.java

```diff
@@ -89,31 +79,39 @@ public class UpdateConverter implements Converter<AttributeUpdateOperation, SubD
   public SubDocumentUpdate convert(
       final AttributeUpdateOperation operation, final RequestContext context)
       throws ConversionException {
-    final AttributeOperationMetadata metadata = OPERATOR_MAP.get(operation.getOperator());
+    final UpdateOperator operator = OPERATOR_MAP.get(operation.getOperator());
 
-    if (metadata == null) {
+    if (operator == null) {
       throw new ConversionException(
```

Comment:
Just trying to understand. 
This is not a validator class. This is a part of the converters. Do we still want to through Status exceptions from here?

Also, what if this converter class was being called by many clients (a few of them catching the custom `ConversionException` for handling the exception scenarios differently and others rethrowing)?
Throwing StatusRuntimeException may make the exception part of this method's signature implicit (and hence the clients may forget to catch runtime exceptions).

Moreover, even if we were throwing Status exception from here, it would be caught in `org.hypertrace.entity.query.service.EntityQueryServiceImpl.bulkUpdateAllMatchingFilter`, which would then wrap this in an `INTERNAL` status exception. So, the clients would anyways be receiving an `INTERNAL` error despite with through Status exception or others from here.

> grpc status with trailers

Sure, I'll add the trailers to the original GRPC handler method.

Reply:
Yeah, fair point - it depends on how it's used. The main goal is that anything returned from the service returns a meaningful error rather than an "internal - unknown" type error. For this, if I'm reading it correctly - doesn't this mean the caller sent an operator that was not supported? So that seems useful to propagate back as an invalid argument.

There may be bigger refactors needed to make that work cleanly and will let you judge if it's in scope or not, but that's the high level point - that the vast majority of exceptions we throw are intended to propagate back, and should be constructed in a form to support that. If it is intended to be caught by the caller, then - dare I say it? - perhaps a checked exception is a better fit (which I now see you're using here). So who is looking for that checked exception, and what are they doing with it?

Reply:
> doesn't this mean the caller sent an operator that was not supported?

That's right! If we had to make this a part of the request validator, it would have to query attribute service. Not sure if it is a good idea to make the validator talk to external services. Moreover, on-the-fly validation designates validating the input to different classes (each one will be responsible for validating and processing). If we had to do that in the validator class, we would have to get the object that could be deeply nested and perform the validation.


I've added a version to capture it and convert it into an INVALID_ARGUMENT. Please take a look and let me know if there's any better way to handle it. Note: I have not done it for all the code paths in bulk update. Did only for the paths I touched in this PR.

Reply:
Basically, the idea is to throw Status exception (but as a checked exception than as a runtime exception).
Any checked exception would be caught by the service implementation method, logged, and notified to the clients.